2016 04 stze result api call

stze opened the pull request:
This PR introduces the result api call. You can test this by using `2016-05-bontric-implement-result-call` python client branch.
- Fixed handle run callid bug
- Implemented result api call + functional test
- Fixed coverage 
- Introduce travis gcc build (used for coverage)

Commit:
Implemented result api call

Commit:
Merge branch 'master' into 2016-04-stze-result-api-call

Commit:
Fixed run api call handler bug and refactored result api call code

Commit:
Try to fix coveralls

Commit:
Disable coverage use when using clang

Commit:
Disable sudo in travis

Commit:
Fix permissions and revert sudo remove

Commit:
Check if response is an error

Commit:
Initialize callinfo error variable

Commit:
Reduce heap usage in connection request handler

stze commented:
@mkind i'll hand over this PR and branch to you. If you'll find some problems, would be great if you could fix them, otherwise please merge into master.

mkind committed:
fixing some compiler warnings by replacing -fno-sanitize-recover w/ -fno-sanitize-recover=undefined,integer

mkind committed:
fixing typo

mkind committed:
adding some verbosity output

mkind committed:
only set error if not already set. this prevents overwritting a more detailed error message

mkind committed:
Refactoring test. The dispatcher tests now make use of central
validation functions to validate sent requests and sent responses. In
addition to that, the test message are all built by the same method. So
no more redundant code in all dispatcher tests.

mkind committed:
Cancel if an error occured during result forwarding.

mkind committed:
verify, that the forwarded callid is correct

mkind committed:
If callid 0x0 is passed, the hashmap returns a targetpluginkey 0x0. This
breaks the khashmap_get in connection_send_request.

mkind committed:
regression test that handles error fixed in commit 1c66e9d

mkind committed:
Revert "fixing some compiler warnings by replacing -fno-sanitize-recover w/ -fno-sanitize-recover=undefined,integer"

This reverts commit 456a6ad8aea1731768fc8982255a2709327d8eb9.

stze commented on CMakeLists.txt:
```diff
@@ -327,13 +329,13 @@ endif()
 if(CLANG_ADDRESS_SANITIZER)
   set_property(TARGET sb APPEND_STRING PROPERTY COMPILE_FLAGS "-DEXITFREE ")
-  set_property(TARGET sb APPEND_STRING PROPERTY COMPILE_FLAGS "-fno-sanitize-recover -fno-omit-frame-pointer -fno-optimize-sibling-calls -fsanitize=address -fsanitize=undefined ")
+  set_property(TARGET sb APPEND_STRING PROPERTY COMPILE_FLAGS "-fno-sanitize-recover=undefined,integer -fno-omit-frame-pointer -fno-optimize-sibling-calls -fsanitize=address -fsanitize=undefined ")
```
this change seems reasonable, but it reverts an old commit where this option was disabled. travis clang is old and doesn't support `undefined,integer` yet.

stze commented on src/api/result.c:
```diff
@@ -68,9 +68,12 @@ int api_result(char *targetpluginkey, uint64_t callid,
   cinfo = connection_send_request(targetpluginkey, result, result_params,
       api_error);
 
+  if (true == api_error->isset)
```
why not:

``` c
if(api_error->isset) 
```

and

``` c
if(!api_error->isset) 
```

instead of 

``` c
if(true == api_error->isset) 
```

and 

``` c
if(false == api_error->isset) 
```

mkind replied:
no problem with your proposal

mkind replied:
Fixed this in 0e6e4bb134237cc2e399ad2a7002f01a39dbe590.

stze commented on test/helper-all.c:
```diff
@@ -63,3 +66,251 @@ void connect_and_create(char *pluginkey)
+  if (!p || !f)
+    LOG_ERROR("[test] Failed to alloc mem for example plugin.\n");
+
+  p->key = cstring_copy_string("0123456789ABCDEF");
```
would be nice to allocate all these strings on the stack. maybe we can change it to:

``` c
p->key = (string) {.str = "0123456789ABCDEF", .length = sizeof("0123456789ABCDEF") - 1}};
```

then we can remove the `free_string` calls in the `helper_free_plugin` function.

mkind replied:
This might be a general approach to save some heap allocs in other contexts too.

mkind replied:
fixed in a30446b5c44b78228accd5fdd030277a75c92f18

stze reviewed (commented):
Looks good to me @mkind. Only a few comments.

mkind committed:
This commit introduces the possibility that a plugin can call other
plugins without providing rpc calls itself. This is reasonable for
different kinds of UI.

mkind committed:
adding missing test

mkind committed:
testing whether params size is evaluated correctly

mkind committed:
removing some heap allocations in the test helper-all
